
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      <link rel="icon" href="../../assets/images/SIB_logo.svg">
      <meta name="generator" content="mkdocs-1.4.1, mkdocs-material-8.5.7">
    
    
      
        <title>Differential gene expression - Single cell transcriptomics</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.20d9efc8.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  


  <script id="__analytics">function __md_analytics(){function n(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],n("js",new Date),n("config","G-1H5B39PY7V"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){this.value&&n("event","search",{search_term:this.value})}),document$.subscribe(function(){var a=document.forms.feedback;if(void 0!==a)for(var e of a.querySelectorAll("[type=submit]"))e.addEventListener("click",function(e){e.preventDefault();var t=document.location.pathname,e=this.getAttribute("data-md-value");n("event","feedback",{page:t,data:e}),a.firstElementChild.disabled=!0;e=a.querySelector(".md-feedback__note [data-md-value='"+e+"']");e&&(e.hidden=!1)}),a.hidden=!1}),location$.subscribe(function(e){n("config","G-1H5B39PY7V",{page_path:e.pathname})})});var e=document.createElement("script");e.async=!0,e.src="https://www.googletagmanager.com/gtag/js?id=G-1H5B39PY7V",document.getElementById("__analytics").insertAdjacentElement("afterEnd",e)}</script>

  
    <script>var consent;"undefined"==typeof __md_analytics||(consent=__md_get("__consent"))&&consent.analytics&&__md_analytics()</script>
  

    
    
    
<meta property="og:url"                content="https://sib-swiss.github.io/single-cell-training/latest/" />
<meta property="og:title"              content="Single cell transcriptomics" />
<meta property="og:description"        content="Course on single cell transcriptomics" />
<meta property="og:image"              content="https://sib-swiss.github.io/single-cell-training/2022.11/assets/images/umap_nonintegrated_dim50.png" />

<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Single cell transcriptomics" />
<meta name="twitter:description" content="Course on single cell transcriptomics" />
<meta name="twitter:image" content="https://sib-swiss.github.io/single-cell-training/2022.11/assets/images/umap_nonintegrated_dim50.png" />

  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#material" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
        <aside class="md-banner">
          <div class="md-banner__inner md-grid md-typeset">
            
              <button class="md-banner__button md-icon" aria-label="Don't show this again">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
              </button>
            
            
  <b>Issues, questions or feedback? &#128591; &#128640; </b> Use the "Questions and Answers" section at the bottom &#128071; of every page

          </div>
          
            <script>var content,el=document.querySelector("[data-md-component=announce]");el&&(content=el.querySelector(".md-typeset"),__md_hash(content.innerHTML)===__md_get("__announce")&&(el.hidden=!0))</script>
          
        </aside>
      
    </div>
    
      <div data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Single cell transcriptomics" class="md-header__button md-logo" aria-label="Single cell transcriptomics" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Single cell transcriptomics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Differential gene expression
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/sib-swiss/single-cell-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/single-cell-training
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Single cell transcriptomics" class="md-nav__button md-logo" aria-label="Single cell transcriptomics" data-md-component="logo">
      
  <img src="../../assets/images/SIB_logo.svg" alt="logo">

    </a>
    Single cell transcriptomics
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/sib-swiss/single-cell-training" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    sib-swiss/single-cell-training
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../precourse/" class="md-nav__link">
        Precourse preparations
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../course_schedule/" class="md-nav__link">
        Course schedule
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      
      
      
        <label class="md-nav__link" for="__nav_4">
          Day 1
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 1" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Day 1
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/setup/" class="md-nav__link">
        Setup
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/introduction_cellranger/" class="md-nav__link">
        Introduction & cellranger
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/analysis_tools_qc/" class="md-nav__link">
        Analysis tools and QC
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day1/normalization_scaling/" class="md-nav__link">
        Normalization and scaling
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_5" type="checkbox" id="__nav_5" >
      
      
      
      
        <label class="md-nav__link" for="__nav_5">
          Day 2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 2" data-md-level="1">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Day 2
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/dimensionality_reduction/" class="md-nav__link">
        Dimensionality reduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/integration/" class="md-nav__link">
        Integration
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/clustering/" class="md-nav__link">
        Clustering
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../day2/cell_annotation/" class="md-nav__link">
        Cell annotation
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_6" type="checkbox" id="__nav_6" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_6">
          Day 3
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Day 3" data-md-level="1">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          Day 3
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Differential gene expression
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Differential gene expression
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#find-all-markers-for-each-cluster" class="md-nav__link">
    Find all markers for each cluster
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#differential-expression-between-groups-of-cells" class="md-nav__link">
    Differential expression between groups of cells
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#differential-expression-using-limma" class="md-nav__link">
    Differential expression using limma
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../enrichment_analysis/" class="md-nav__link">
        Enrichment analysis
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../advanced_analyses/" class="md-nav__link">
        Advanced analyses
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../trajectory_analysis/" class="md-nav__link">
        Trajectory analysis (extra)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#material" class="md-nav__link">
    Material
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exercises" class="md-nav__link">
    Exercises
  </a>
  
    <nav class="md-nav" aria-label="Exercises">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#find-all-markers-for-each-cluster" class="md-nav__link">
    Find all markers for each cluster
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#differential-expression-between-groups-of-cells" class="md-nav__link">
    Differential expression between groups of cells
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#differential-expression-using-limma" class="md-nav__link">
    Differential expression using limma
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                

                  

  <a href="https://github.com/sib-swiss/single-cell-training/edit/master/docs/day3/differential_gene_expression.md" title="Edit this page" class="md-content__button md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>


  <h1>Differential gene expression</h1>

<h2 id="material">Material</h2>
<p><a class="md-button" href="../../assets/pdf/DGE_and_enrichment_analysis.pdf"><span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M64 0C28.7 0 0 28.7 0 64v384c0 35.3 28.7 64 64 64h256c35.3 0 64-28.7 64-64V160H256c-17.7 0-32-14.3-32-32V0H64zm192 0v128h128L256 0zM64 224h24c30.9 0 56 25.1 56 56s-25.1 56-56 56h-8v32c0 8.8-7.2 16-16 16s-16-7.2-16-16V240c0-8.8 7.2-16 16-16zm24 80c13.3 0 24-10.7 24-24s-10.7-24-24-24h-8v48h8zm72-64c0-8.8 7.2-16 16-16h24c26.5 0 48 21.5 48 48v64c0 26.5-21.5 48-48 48h-24c-8.8 0-16-7.2-16-16V240zm32 112h8c8.8 0 16-7.2 16-16v-64c0-8.8-7.2-16-16-16h-8v96zm96-128h48c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v32h32c8.8 0 16 7.2 16 16s-7.2 16-16 16h-32v48c0 8.8-7.2 16-16 16s-16-7.2-16-16V240c0-8.8 7.2-16 16-16z"/></svg></span> Download the presentation</a></p>
<ul>
<li>More information on <a href="https://bioconductor.org/books/release/OSCA/multi-sample-comparisons.html#differential-expression-between-condition">pseudobulk analysis</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/muscat/inst/doc/analysis.html">Muscat</a> for pseudobulk DGE.</li>
<li><a href="https://www.nature.com/articles/nmeth.4612">Paper</a> on the robustness of different differential expression analysis methods</li>
</ul>
<h2 id="exercises">Exercises</h2>
<h3 id="find-all-markers-for-each-cluster">Find all markers for each cluster</h3>
<p>Load the <code>seu_int</code> dataset you have created yesterday:</p>
<div class="highlight"><pre><span></span><code><span class="n">seu_int</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;seu_int_day2_part2.rds&quot;</span><span class="p">)</span>
</code></pre></div>
<p>And load the following packages (install them if they are missing):</p>
<div class="highlight"><pre><span></span><code><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">edgeR</span><span class="p">)</span><span class="w"> </span><span class="c1"># BiocManager::install(&quot;edgeR&quot;)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">limma</span><span class="p">)</span>
</code></pre></div>
<p>The function <code>FindAllMarkers</code> performs a Wilcoxon plot to determine the genes differentially expressed between each cluster and the rest of the cells. Other types of tests than the Wilcoxon test are available. Check it out by running <code>?Seurat::FindAllMarkers</code>.</p>
<p>Now run analysis:</p>
<div class="highlight"><pre><span></span><code><span class="n">de_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Seurat</span><span class="o">::</span><span class="nf">FindAllMarkers</span><span class="p">(</span><span class="n">seu_int</span><span class="p">,</span><span class="w">  </span><span class="n">min.pct</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.25</span><span class="p">,</span>
<span class="w">                                   </span><span class="n">only.pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Time for coffee</p>
<p>This takes a while. Have a break.</p>
</div>
<p>Subset the table to only keep the significant genes, and you can save it as a csv file if you wish to explore it further. Then extract the top 3 markers per cluster:</p>
<div class="highlight"><pre><span></span><code><span class="n">de_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">de_genes</span><span class="p">,</span><span class="w"> </span><span class="n">de_genes</span><span class="o">$</span><span class="n">p_val_adj</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">)</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">de_genes</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;de_genes_FindAllMarkers.csv&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">F</span><span class="p">)</span>


<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="n">top_specific_markers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">de_genes</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">top_n</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">avg_log2FC</span><span class="p">)</span>
</code></pre></div>
<p>And generate e.g. a dotplot:</p>
<div class="highlight"><pre><span></span><code><span class="n">dittoSeq</span><span class="o">::</span><span class="nf">dittoDotPlot</span><span class="p">(</span><span class="n">seu_int</span><span class="p">,</span><span class="w"> </span><span class="n">vars</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">top_specific_markers</span><span class="o">$</span><span class="n">gene</span><span class="p">),</span><span class="w"> </span>
<span class="w">                       </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;integrated_snn_res.0.3&quot;</span><span class="p">)</span>
</code></pre></div>
<figure>
    <img src="../../assets/images/dotplot_degenes.png" width="700"/>
</figure>

<p><strong>Exercise:</strong> What are significant marker genes in cluster 0 and 8? Are the T cell genes in there?</p>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>You can re-load the vector with immune genes with:</p>
<div class="highlight"><pre><span></span><code><span class="n">tcell_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;IL7R&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;LTB&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;TRAC&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD3D&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
<details class="done">
<summary>Answer</summary>
<p>Running</p>
<div class="highlight"><pre><span></span><code><span class="n">de_genes</span><span class="p">[</span><span class="n">de_genes</span><span class="o">$</span><span class="n">gene</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">tcell_genes</span><span class="p">,]</span>
</code></pre></div>
<p>Returns:
<div class="highlight"><pre><span></span><code>               p_val avg_log2FC pct.1 pct.2     p_val_adj cluster gene
CD3D    0.000000e+00  2.0432771 0.768 0.228  0.000000e+00       0 CD3D
TRAC   1.157793e-287  1.6805543 0.618 0.205 2.161947e-283       0 TRAC
LTB    1.072643e-266  1.5215326 0.757 0.395 2.002946e-262       0  LTB
IL7R   5.112493e-211  1.5236657 0.438 0.114 9.546557e-207       0 IL7R
LTB.7   6.554548e-36  1.1405474 0.672 0.465  1.223931e-31       7  LTB
TRAC.8 7.568418e-117  1.8472689 0.759 0.273 1.413251e-112       8 TRAC
CD3D.8 3.079377e-110  1.7144870 0.800 0.326 5.750121e-106       8 CD3D
LTB.8   1.580808e-61  1.6529117 0.774 0.461  2.951843e-57       8  LTB
IL7R.2  4.497526e-45  1.1489439 0.458 0.173  8.398231e-41       8 IL7R
LTB.11  2.727014e-25  0.8193337 0.750 0.467  5.092153e-21      11  LTB
</code></pre></div></p>
<p>So, yes, the T-cell genes are highly significant markers for cluster 0 and 8.</p>
</details>
<h3 id="differential-expression-between-groups-of-cells">Differential expression between groups of cells</h3>
<p>The <code>FindMarkers</code> function allows to test for differential gene expression analysis specifically between 2 groups of cells, i.e. perform pairwise comparisons, eg between cells of cluster 0 vs cluster 2, or between cells annotated as T-cells and B-cells.</p>
<p>First we can set the default cell identity to the cell types defined by <code>SingleR</code>:</p>
<div class="highlight"><pre><span></span><code><span class="n">seu_int</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Seurat</span><span class="o">::</span><span class="nf">SetIdent</span><span class="p">(</span><span class="n">seu_int</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;SingleR_annot&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Run the differential gene expression analysis and subset the table to keep the significant genes:</p>
<div class="highlight"><pre><span></span><code><span class="n">deg_cd8_cd4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Seurat</span><span class="o">::</span><span class="nf">FindMarkers</span><span class="p">(</span><span class="n">seu_int</span><span class="p">,</span>
<span class="w">                                   </span><span class="n">ident.1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD8+ T cells&quot;</span><span class="p">,</span>
<span class="w">                                   </span><span class="n">ident.2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;CD4+ T cells&quot;</span><span class="p">,</span>
<span class="w">                                   </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seu_int</span><span class="o">$</span><span class="n">SingleR_annot</span><span class="p">,</span>
<span class="w">                                   </span><span class="n">test.use</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;wilcox&quot;</span><span class="p">)</span>
<span class="n">deg_cd8_cd4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">deg_cd8_cd4</span><span class="p">,</span><span class="w"> </span><span class="n">deg_cd8_cd4</span><span class="o">$</span><span class="n">p_val_adj</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">)</span>
</code></pre></div>
<p><strong>Exercise:</strong> Are CD8A, CD8B and CD4 in there? What does the sign (i.e. positive or negative) mean in the log fold change values? Are they according to the CD8+ and CD4+ annotations? Check your answer by generating a violin plot of a top differentially expressed gene.</p>
<details class="done">
<summary>Answer</summary>
<p>You can check out the results with:</p>
<div class="highlight"><pre><span></span><code><span class="nf">View</span><span class="p">(</span><span class="n">deg_cd8_cd4</span><span class="p">)</span>
</code></pre></div>
<p>For an explanation of the log fold change have a look at <code>?Seurat::FindMarkers</code>. At <strong>Value</strong> it says:</p>
<blockquote>
<p><code>avg_logFC</code>: log fold-chage of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group</p>
</blockquote>
<p>To view CD8A, CD8B and CD4:</p>
<div class="highlight"><pre><span></span><code><span class="n">deg_cd8_cd4</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD4&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8A&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8B&quot;</span><span class="p">),]</span>
</code></pre></div>
<p>Returning:</p>
<div class="highlight"><pre><span></span><code>            p_val avg_log2FC pct.1 pct.2    p_val_adj
CD4  1.070126e-13 -0.4000835 0.012 0.103 1.998246e-09
CD8A 1.409306e-77  1.2956354 0.344 0.008 2.631597e-73
CD8B 7.113148e-36  0.8536693 0.479 0.177 1.328238e-31
</code></pre></div>
<p>Indeed, because we compared ident.1 = &ldquo;CD8+ T cells&rdquo; to ident.2 = &ldquo;CD4+ T cells&rdquo;, a negative log2FC for the CD4 gene indicates a lower expression in CD8+ T-cells than in CD4+ T-cells, while a positive log2FC for the CD8A and CD8B genes indicates a higher expression in CD8+ T-cells.</p>
<p>Plotting the genes in these two T-cell groups only:</p>
<div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">VlnPlot</span><span class="p">(</span><span class="n">seu_int</span><span class="p">,</span><span class="w"> </span>
<span class="w">            </span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD4&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8A&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD8B&quot;</span><span class="p">),</span>
<span class="w">            </span><span class="n">idents</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;CD8+ T cells&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;CD4+ T cells&quot;</span><span class="p">))</span>
</code></pre></div>
<p>Returning:</p>
<p><figure>
  <img src="../../assets/images/violinplot_CD8_CD4.png" width="600"/>
</figure></p>
</details>
<h3 id="differential-expression-using-limma">Differential expression using <code>limma</code></h3>
<p>The Wilcoxon test implemented in <code>FindMarkers</code> does not allow you to test for complex design (eg factorial experiments) or to include batch as a covariate. It doesn&rsquo;t allow you to run paired-sample T tests for example. </p>
<p>For more complex designs, we can use <code>edgeR</code> or <code>limma</code> which are designed for microarray or bulk RNA seq data and provide a design matrix that includes covariates for example, or sample IDs for paired analyses.</p>
<p>We will load an object containing only pro B cells, both from healthy tissues (PBMMC), and malignant tissues (ETV6-RUNX1). </p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Please NOTE that in the original design of this data set, the healthy and malignant tissues were not  patient-matched, i.e. the real design was not the one of paired healthy and malignant tissues. However, for demonstration purposes, we will show you how to run a paired analysis, and do as if the PBMMC-1 and ETV6-RUNX1-1 samples both came from the same patient 1, the PBMMC-2 and ETV6-RUNX1-2 samples both came from the same patient 2, etc&hellip;</p>
</div>
<p>We can load the object and explore its UMAP and meta.data like this:</p>
<div class="highlight"><pre><span></span><code><span class="n">proB</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readRDS</span><span class="p">(</span><span class="s">&quot;course_data/proB.rds&quot;</span><span class="p">)</span>

<span class="nf">DimPlot</span><span class="p">(</span><span class="n">proB</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;orig.ident&quot;</span><span class="p">)</span>

<span class="nf">table</span><span class="p">(</span><span class="n">proB</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">type</span><span class="p">)</span>
<span class="c1"># ETV6-RUNX1      PBMMC </span>
<span class="c1">#      2000       1021</span>

<span class="nf">head</span><span class="p">(</span><span class="n">proB</span><span class="o">@</span><span class="n">meta.data</span><span class="p">)</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you want to know how this pro-B cell subset is generated, have a look at the script <a href="../../assets/scripts/generate_object_proB.R">here</a>.</p>
</div>
<p>Since we will start with differential gene expression, we set the default assay back to &ldquo;RNA&rdquo;. Also, we set the default identity to the cell type:</p>
<div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">proB</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;RNA&quot;</span>
<span class="n">Seurat</span><span class="o">::</span><span class="nf">Idents</span><span class="p">(</span><span class="n">proB</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">proB</span><span class="o">$</span><span class="n">orig.ident</span>
</code></pre></div>
<p>Let&rsquo;s have a look at the UMAP (again), coloured by celltype:</p>
<div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">proB</span><span class="p">)</span>
</code></pre></div>
<p>Let&rsquo;s say we are specifically interested to test for differential gene expression between the tumor and normal samples.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Here we could also test for e.g. healthy versus diseased within a celltype/cluster.</p>
</div>
<p>Now we will run differential expression analysis between tumor and healthy cells using the patient ID as a covariate by using <code>limma</code>.</p>
<p>Prepare the pseudobulk count matrix:</p>
<div class="highlight"><pre><span></span><code><span class="c1">#taking the proB data </span>
<span class="n">Seurat</span><span class="o">::</span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">proB</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;RNA&quot;</span>
<span class="n">Seurat</span><span class="o">::</span><span class="nf">Idents</span><span class="p">(</span><span class="n">proB</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">proB</span><span class="o">$</span><span class="n">orig.ident</span>

<span class="c1">## add the patient id also for paired DGE</span>
<span class="n">proB</span><span class="o">$</span><span class="n">patient.id</span><span class="o">&lt;-</span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;ETV6-RUNX1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ETV6_RUNX1&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">proB</span><span class="o">$</span><span class="n">orig.ident</span><span class="p">)</span>
<span class="n">proB</span><span class="o">$</span><span class="n">patient.id</span><span class="o">&lt;-</span><span class="nf">sapply</span><span class="p">(</span><span class="nf">strsplit</span><span class="p">(</span><span class="n">proB</span><span class="o">$</span><span class="n">patient.id</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;-&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&#39;[&#39;</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="c1">## Here we do perform pseudo-bulk:</span>
<span class="c1">##first a mandatory column of sample needs to be added to the meta data that is the grouping factor, should be the samples</span>
<span class="n">proB</span><span class="o">$</span><span class="n">sample</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="n">proB</span><span class="o">$</span><span class="n">orig.ident</span><span class="p">)</span>

<span class="c1">##first an sce object is needed</span>
<span class="n">sce_proB</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.SingleCellExperiment</span><span class="p">(</span><span class="n">proB</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">scuttle</span><span class="p">)</span>

<span class="c1">##aggregateAcrossCells here it is only aggregated by sample, one could imagine</span>
<span class="c1">##to aggregate by sample and by celltype for instance</span>
<span class="n">summed</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">aggregateAcrossCells</span><span class="p">(</span><span class="n">sce_proB</span><span class="p">,</span><span class="w"> </span>
<span class="w">                               </span><span class="n">id</span><span class="o">=</span><span class="nf">colData</span><span class="p">(</span><span class="n">sce_proB</span><span class="p">)[,</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;sample&quot;</span><span class="p">)])</span>

<span class="c1">##have a look at the counts</span>
<span class="nf">counts</span><span class="p">(</span><span class="n">summed</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">,]</span>

<span class="c1">#have a look at the colData of our new object summed, can you see type and </span>
<span class="c1">#patient.id are there</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">colData</span><span class="p">(</span><span class="n">summed</span><span class="p">))</span>
</code></pre></div>
<p>Generate a <code>DGEList</code> object to use as input for <code>limma</code> and filter the genes to remove lowly expressed genes. How many are left?</p>
<div class="highlight"><pre><span></span><code><span class="c1">#As in the standard limma analysis generate a DGE object</span>

<span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DGEList</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">summed</span><span class="p">),</span><span class="w"> </span><span class="n">samples</span><span class="o">=</span><span class="nf">colData</span><span class="p">(</span><span class="n">summed</span><span class="p">)</span><span class="o">$</span><span class="n">sample</span><span class="p">)</span>

<span class="c1">##filter lowly expressed (recommanded for limma)</span>
<span class="n">keep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">filterByExpr</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="o">=</span><span class="n">summed</span><span class="o">$</span><span class="n">type</span><span class="p">)</span>
<span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">y</span><span class="p">[</span><span class="n">keep</span><span class="p">,]</span>

<span class="c1">##see how many genes were kept </span>
<span class="nf">summary</span><span class="p">(</span><span class="n">keep</span><span class="p">)</span>
</code></pre></div>
<p>Generate a design matrix, including patient ID to model for a paired analysis. If you need help to generate a design matrix, check out the very nice <a href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeR User Guide</a>, sections 3.3 and 3.4.
Extract the sample ID from the meta.data, then create the design matrix:</p>
<div class="highlight"><pre><span></span><code><span class="c1">## Create the design matrix and include the technology as a covariate:</span>
<span class="n">design</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="m">0</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">summed</span><span class="o">$</span><span class="n">type</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">summed</span><span class="o">$</span><span class="n">patient.id</span><span class="p">)</span>

<span class="c1"># Have a look</span>
<span class="n">design</span>

<span class="c1"># change column/rownames names to more simple group names: </span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">design</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make.names</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ETV6-RUNX1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;PBMMC&quot;</span><span class="p">,</span><span class="s">&quot;patient2&quot;</span><span class="p">,</span><span class="s">&quot;patient3&quot;</span><span class="p">))</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">design</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">colData</span><span class="p">(</span><span class="n">summed</span><span class="p">)</span><span class="o">$</span><span class="n">sample</span>
</code></pre></div>
<p>Specify which contrast to analyse:</p>
<div class="highlight"><pre><span></span><code><span class="n">contrast.mat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">limma</span><span class="o">::</span><span class="nf">makeContrasts</span><span class="p">(</span><span class="n">ETV6.RUNX1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">PBMMC</span><span class="p">,</span>
<span class="w">                                     </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">design</span><span class="p">)</span>
</code></pre></div>
<p>Firt, we perform TMM normalization using edgeR, and then <code>limma</code> can perform the transformation with <code>voom</code>, fit the model, compute the contrasts and compute test statistics with <code>eBayes</code>:</p>
<div class="highlight"><pre><span></span><code><span class="n">dge</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">edgeR</span><span class="o">::</span><span class="nf">calcNormFactors</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="w">  </span>

<span class="c1">#Do limma</span>
<span class="n">vm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">limma</span><span class="o">::</span><span class="nf">voom</span><span class="p">(</span><span class="n">dge</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">design</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">limma</span><span class="o">::</span><span class="nf">lmFit</span><span class="p">(</span><span class="n">vm</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">design</span><span class="p">)</span>
<span class="n">fit.contrasts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">limma</span><span class="o">::</span><span class="nf">contrasts.fit</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span><span class="w"> </span><span class="n">contrast.mat</span><span class="p">)</span>
<span class="n">fit.contrasts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">limma</span><span class="o">::</span><span class="nf">eBayes</span><span class="p">(</span><span class="n">fit.contrasts</span><span class="p">)</span>
</code></pre></div>
<p>We can use <code>topTable</code> to get the most significantly differentially expressed genes, and save the full DE results to an object. How many genes are significant? Are you suprised by this number?</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Show the top differentially expressed genes:</span>
<span class="n">limma</span><span class="o">::</span><span class="nf">topTable</span><span class="p">(</span><span class="n">fit.contrasts</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">sort.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;P&quot;</span><span class="p">)</span>
<span class="n">limma_de</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">limma</span><span class="o">::</span><span class="nf">topTable</span><span class="p">(</span><span class="n">fit.contrasts</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">sort.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;P&quot;</span><span class="p">)</span>
<span class="nf">length</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">limma_de</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">))</span>
</code></pre></div>
<p>And we can check whether this corresponds to the counts by generating a violin plot, or a gene downregulated in tumor, or a gene upregulated in tumor:</p>
<div class="highlight"><pre><span></span><code><span class="n">Seurat</span><span class="o">::</span><span class="nf">VlnPlot</span><span class="p">(</span><span class="n">proB</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;S100A9&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;type&quot;</span><span class="p">)</span>
<span class="n">Seurat</span><span class="o">::</span><span class="nf">VlnPlot</span><span class="p">(</span><span class="n">proB</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;SOCS2&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">split.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;type&quot;</span><span class="p">)</span>
</code></pre></div>
<p>We can run a similar analysis with <code>Seurat</code>, but this will not take into account the paired design. Run the code below. </p>
<div class="highlight"><pre><span></span><code><span class="n">tum_vs_norm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Seurat</span><span class="o">::</span><span class="nf">FindMarkers</span><span class="p">(</span><span class="n">proB</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">ident.1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;ETV6-RUNX1&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">ident.2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;PBMMC&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;type&quot;</span><span class="p">)</span>
<span class="n">tum_vs_norm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">subset</span><span class="p">(</span><span class="n">tum_vs_norm</span><span class="p">,</span><span class="w"> </span><span class="n">tum_vs_norm</span><span class="o">$</span><span class="n">p_val_adj</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">)</span>
</code></pre></div>
<p><strong>Exercise (extra):</strong> How many genes are significant? How does the fold change of these genes compare to the fold change of the top genes found by limma?</p>
<details class="done">
<summary>Answer</summary>
<p><div class="highlight"><pre><span></span><code><span class="nf">dim</span><span class="p">(</span><span class="n">tum_vs_norm</span><span class="p">)</span><span class="w"> </span>
</code></pre></div>
We find 1893 significant genes. If we merge the <code>FindMarkers</code> and the <code>limma</code> results, keep <code>limma</code>&lsquo;s most significant genes and plot:
<div class="highlight"><pre><span></span><code><span class="n">merge_limma_FindMarkers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">merge</span><span class="p">(</span><span class="n">tum_vs_norm</span><span class="p">,</span><span class="w"> </span><span class="n">limma_de</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s">&quot;row.names&quot;</span><span class="p">,</span>
<span class="w">                           </span><span class="n">all.x</span><span class="o">=</span><span class="bp">T</span><span class="p">)</span>

<span class="nf">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="m">4</span><span class="p">,</span><span class="m">4</span><span class="p">,</span><span class="m">4</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">merge_limma_FindMarkers</span><span class="o">$</span><span class="n">avg_log2FC</span><span class="p">,</span>
<span class="w">    </span><span class="n">merge_limma_FindMarkers</span><span class="o">$</span><span class="n">logFC</span><span class="p">,</span>
<span class="w">    </span><span class="n">xlab</span><span class="o">=</span><span class="s">&quot;log2FC Wilcoxon&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s">&quot;log2FC limma&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">pch</span><span class="o">=</span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">b</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">)</span>
</code></pre></div>
Returning:</p>
<p><figure>
  <img src="../../assets/images/limma_vs_wilcoxon.png" width="600"/>
</figure></p>
</details>
<div class="admonition warning">
<p class="admonition-title">keep the object</p>
<p>Keep the <code>tum_vs_norm</code> object because we will use this output object for the enrichment analysis in the next section.</p>
</div>


  




                



<!-- Giscus -->
<!-- <h2 id="__comments">Comments</h2> -->
<h2>Questions & Answers</h2>

<script src="https://giscus.app/client.js"
        data-repo="sib-swiss/single-cell-training"
        data-repo-id="MDEwOlJlcG9zaXRvcnkzNzA1ODY2OTU="
        data-category="Q&A"
        data-category-id="DIC_kwDOFha0R84CQ3Gn"
        data-mapping="title"
        data-strict="0"
        data-reactions-enabled="1"
        data-emit-metadata="0"
        data-input-position="top"
        data-theme="preferred_color_scheme"
        data-lang="en"
        crossorigin="anonymous"
        async>
</script>

<!-- Synchronize Giscus theme with palette -->
<script>
    var giscus = document.querySelector("script[src*=giscus]")

    /* Set palette on initial load */
    var palette = __md_get("__palette")
    if (palette && typeof palette.color === "object") {
        var theme = palette.color.scheme === "slate" ? "dark" : "light"
        giscus.setAttribute("data-theme", theme)
    }

    /* Register event handlers after documented loaded */
    document.addEventListener("DOMContentLoaded", function () {
        var ref = document.querySelector("[data-md-component=palette]")
        ref.addEventListener("change", function () {
            var palette = __md_get("__palette")
            if (palette && typeof palette.color === "object") {
                var theme = palette.color.scheme === "slate" ? "dark" : "light"

                /* Instruct Giscus to change theme */
                var frame = document.querySelector(".giscus-frame")
                frame.contentWindow.postMessage(
                    { giscus: { setConfig: { theme } } },
                    "https://giscus.app"
                )
            }
        })
    })
</script>

              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../../day2/cell_annotation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Cell annotation" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Cell annotation
            </div>
          </div>
        </a>
      
      
        
        <a href="../enrichment_analysis/" class="md-footer__link md-footer__link--next" aria-label="Next: Enrichment analysis" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Enrichment analysis
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2021 - 2022 SIB Swiss Institute of Bioinformatics – <a href="#__consent">Change cookie settings</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-consent" data-md-component="consent" id="__consent" hidden>
        <div class="md-consent__overlay"></div>
        <aside class="md-consent__inner">
          <form class="md-consent__form md-grid md-typeset" name="consent">
            

  
    
  


  
    
  



  

<h4>Cookie consent</h4>
<p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our documentation better.</p>
<input type="checkbox" class="md-toggle" id="__settings">
<div class="md-consent__settings">
  <ul class="task-list">
    
      
        
        
      
      <li class="task-list-item">
        <label class="task-list-control">
          <input type="checkbox" name="analytics" checked>
          <span class="task-list-indicator"></span>
          Google Analytics
        <label>
      </li>
    
      
        
        
      
      <li class="task-list-item">
        <label class="task-list-control">
          <input type="checkbox" name="github" checked>
          <span class="task-list-indicator"></span>
          GitHub
        <label>
      </li>
    
  </ul>
</div>
<div class="md-consent__controls">
  
    
      <button class="md-button md-button--primary">Accept</button>
    
    
    
  
    
    
    
      <label class="md-button" for="__settings">Manage settings</label>
    
  
</div>
          </form>
        </aside>
      </div>
      <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout(function(){document.querySelector("[data-md-component=consent]").hidden=!1},250);var action,form=document.forms.consent;for(action of["submit","reset"])form.addEventListener(action,function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);console.log(new FormData(form)),__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map(function(e){return[e,!0]}))),location.hash="",location.reload()})</script>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss"], "search": "../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.8492ddcf.min.js"></script>
      
    
  </body>
</html>